### PR TITLE
Fix fetchzip

### LINKED ISSUE #5851: nix-prefetch-zip fails moving files up
cc @cstrahan.

Moving files one directory up (https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/fetchzip/nix-prefetch-zip#L130) may fail if one of the files has the same name as the parent directory.

E.g., applied to http://www.lix.polytechnique.fr/coq/pylons/contribs/files/ABP/v8.4/ABP.tar.gz, nix-prefetch-zip fails with:
mv: cannot move ‘/tmp/tmp.iF77nPUttb/unpacked/ABP.tar.gz/ABP/ABP’ to ‘/tmp/tmp.iF77nPUttb/unpacked/ABP.tar.gz/ABP’: Directory not empty

### PR DESCRIPTION
**Pull request by cstrahan**
This fixes:
- Passing stripRoot.
- Archives containing a single file.
- Archives where the root folder has the same name as one of it's children.

Fixes #5851

@edolstra would you mind reviewing this for me, please? I've tested this thoroughly, but wanted to get your feedback first.

### CODE REVIEW

File: pkgs/build-support/fetchzip/default.nix [hunk 1]
```diff
@@ -23,22 +23,31 @@ lib.overrideDerivation (fetchurl ({
...
-    ''
-    # FIXME: handle zip files that contain a single regular file.
-    + lib.optionalString stripRoot ''
+
       shopt -s dotglob
-      if [ "$(ls -d $out/* | wc -l)" != 1 ]; then
-        echo "error: zip file must contain a single directory."
+    ''
+    + (if stripRoot then ''
+      if [ $(find "$unpackDir" -maxdepth 1 -not -path "$unpackDir" | wc -l) != 1 ]; then
```
Comment (wmertens) [inline pkgs/build-support/fetchzip/default.nix]: you could also do an ls of $unpackDir and do a `wc -l` on it

**Comment by edolstra** [comment]
Looks good to me.

File: pkgs/build-support/fetchzip/default.nix (hunk 1 above)
Comment (vcunat) [inline pkgs/build-support/fetchzip/default.nix]: That was so before, but it was changed to ``find``. Perhaps there is a reason to prefer it.

Comment (wmertens) [inline pkgs/build-support/fetchzip/default.nix]: Oh right, I'm using the github split diff view now and I'm not used to it yet.

the previous version is also a bit weird with the `ls -d $out/*`. I think `ls "$unpackDir"` should work fine.

Curious why the switch...

Comment (cstrahan) [inline pkgs/build-support/fetchzip/default.nix]: Yes, `ls "$unpackDir"` would work fine. I think I was in front of a computer for too long when I made that change; I'll make the switch.

**Comment by linquize** [comment]
Why pull request closed?

**Comment by vcunat** [comment]
Because it was pushed to master in fc013537 and parent. (The ref of this PR got updated to that ancestor of current master, so it may look a little confusing.)

**Comment by linquize** [comment]
ic

**Comment by cstrahan** [comment]
fc01353  introduced a regression. See #5954 for more details, along with the commit that fixed it.

**Pull request closed without merging**